### DISCUSSION TITLE
sequence embedding API with protein sequence or UniProt ID

### DISCUSSION THREAD

**Discussion by ratthachat** [facebookresearch/esm#549] [ESM Metagenomic Atlas]
Hi ESM team, first of all, thanks for the great work!!

I have a question : I would like to use sequence embedding API as shown in:
https://esmatlas.com/about

However, the example requires MGnify ID while I have only protein sequence and Uniprot ID.
Is it possible to map the protein sequence or Uniprot ID to MGnify ID?
(I searched but could not find anywhere how to map from Uniprot ID to MGnify ID)

Thank you very much!!

**Comment by naailkhan28** [comment]
The Atlas is made up of sequences from the MGnify Protein DB which is separate to UniProt - you could search your sequence against MGnify [here](https://www.ebi.ac.uk/metagenomics/sequence-search/search/phmmer) and then find the embedding of the nearest homolog, or you could use the examples from this GitHub repo to generate embeddings from the sequence yourself using Python

**Reply by ratthachat** [reply]
@naailkhan28 Thanks so much for the link! 
About generating the embedding by myself, as I will use the embedding in the free Huggingface space web-app, it doesn't have enough resource to use ESM-3b in the prediction pipeline. Therefore, I need to call an API instead.

The link you provided will be very useful, thanks again!

**Reply by naailkhan28** [reply]
If resources are a problem you can run the model in Google Colab yourself and use some of the tips from [here](https://huggingface.co/docs/accelerate/v0.11.0/en/big_modeling) or [here](https://huggingface.co/docs/transformers/perf_infer_gpu_one) as well, of course speed will take a hit but if you don't have too many sequences it could be worth a try

**Reply by ratthachat** [reply]
@naailkhan28 Sorry to ask one more question, this question is about the MGnify search link you gave above ...
From the link above, I use the default protein sequence and get a lot of homolog returns from MGnify search.

Nevertheless, none of the returned MGnify ID work with ESM-embed API, (The API server returns `b'forbidden'`)
do you happen to know why?

Hi, I use the official python API example 

```
import requests
import numpy as np

# mgnify_id = "MGYP003323900000"  # This is ESM API official example --> always work
# mgnify_id = "MGYP000160620212" # This is a homolog ID return from the MGnify default sequence --> b'forbidden'
mgnify_id = "MGYP000006146753"  # This is another homolog ID return from the MGnify default sequence --> b'forbidden'

url = f"https://api.esmatlas.com/fetchEmbedding/ESM2/{mgnify_id}.bin"
header = {"Accept": "application/octet-stream"}
response = requests.get(url, headers=header)

array = np.frombuffer(response.content, dtype=np.float16)
print(array.shape)
```
